internal/provider/waypoint/data_source_waypoint_action.go

```diff
@@ -0,0 +1,207 @@
+				Optional:    true,
+				Computed:    true,
+			},
+			"description": schema.StringAttribute{
+				Description: "A description of the Action.",
+				Computed:    true,
+			},
+			"request": schema.SingleNestedAttribute{
+				Description: "The kind of HTTP request this should trigger.",
+				Computed:    true,
+				Attributes: map[string]schema.Attribute{
+					"custom": schema.SingleNestedAttribute{
+						Description: "Custom mode allows users to define the HTTP method, the request body, etc.",
+						Computed:    true,
+						Attributes: map[string]schema.Attribute{
+							"method": schema.StringAttribute{
+								Description: "The HTTP method to use for the request.",
+								Computed:    true,
+							},
+							"headers": schema.MapAttribute{
+								Description: "Key value headers to send with the request.",
+								Computed:    true,
+								ElementType: types.StringType,
+							},
+							"url": schema.StringAttribute{
+								Description: "The full URL this request should make when invoked.",
+								Computed:    true,
+							},
+							"body": schema.StringAttribute{
+								Description: "The body to be submitted with the request.",
+								Computed:    true,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *DataSourceAction) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	d.client = client
+}
+
+func (d *DataSourceAction) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data ActionResourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	client := d.client
+	if d.client == nil {
+		resp.Diagnostics.AddError(
+			"Unconfigured HCP Client",
+			"Expected configured HCP client. Please report this issue to the provider developers.",
+		)
+		return
+	}
+
+	projectID := client.Config.ProjectID
+	if !data.ProjectID.IsNull() {
+		projectID = data.ProjectID.ValueString()
+	}
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      projectID,
+	}
+
+	var actionModel *waypoint_models.HashicorpCloudWaypointActionConfig
+	var err error
+
+	actionModel, err = clients.GetAction(ctx, client, loc, data.ID.ValueString(), data.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(err.Error(), "Failed to find action by ID or name")
+		return
+	}
+
+	if actionModel.ID != "" {
+		data.ID = types.StringValue(actionModel.ID)
+	}
+	if actionModel.Name != "" {
+		data.Name = types.StringValue(actionModel.Name)
+	}
+	if actionModel.Description != "" {
```

Comment:
I don't think we should do these checks here, instead we should just set the value to what the API gives us. For example (and hypothetical) if the `Description` were to be cleared out, this pattern would prevent the data source from updating it's value in state. As a rule we should just set the state to reflect the API

Reply:
Makes sense to me, I will update how we are handling description those fields here so they can potentially be empty.